[REVIEW] Fast path when possible for non numeric aggregation

VibhuJawa opened the pull request:
This PR adds fast path when possible for non numeric aggregation for Min/Max . The PR speeds up aggregations by avoiding the apply based path [here](https://github.com/dask-contrib/dask-sql/blob/4d5f7dd4c133f7f2c584db2492c8f0ef3cba7e31/dask_sql/physical/rel/logical/aggregate.py#L31-L44). 
 

### Motivating Pandas Example 
```python
%%time
df =  pd.DataFrame({'grp_col': np.arange(0,100_000).repeat(3),
                    'p_date': [np.datetime64('2019-08', 'D'),np.datetime64('2019-07', 'D'),np.datetime64('2018-07', 'D')]*100_000,
                     })

df = dd.from_pandas(df, npartitions=4)
dc.create_table('table_1', df)

query = '''
        select 
            MIN(p_date) as f_date
        from 
            table_1
        group by
            grp_col
        '''
res = dc.sql(query)
res.compute()
```
### Mainline
```python
CPU times: user 39.5 s, sys: 451 ms, total: 39.9 s
Wall time: 37.1 s
```

### With this PR (236):
```python
CPU times: user 3.84 s, sys: 180 ms, total: 4.02 s
Wall time: 1.58 s
```

VibhuJawa committed:
added extra check for datetime

VibhuJawa committed:
cleaned_up_code

VibhuJawa committed:
added check for string

VibhuJawa committed:
added check for cudfDtype and PandasDtype

VibhuJawa committed:
fixed preference for native vs custom agg

VibhuJawa committed:
removed first from agg

VibhuJawa committed:
added StringDtype to test_compatibility.py

VibhuJawa committed:
added better comment for why we take a different code path with StringsDtype

VibhuJawa committed:
added test for datetime

VibhuJawa committed:
changed native to built in

VibhuJawa committed:
made docstring cleaner

ayushdg commented on dask_sql/physical/rel/logical/aggregate.py:
```diff
@@ -7,6 +7,13 @@
 import pandas as pd
 
+try:
+    import cudf
```
unused import

VibhuJawa replied:
Resolved with https://github.com/rapidsai/dask-sql/pull/3/commits/8c8c305f2399e1d368a307d754a8f77f5c422ec5

ayushdg reviewed (changes requested):
Thanks! Just added a small comment on an unused import.

Could you also retarget the PR to target branch-21.12? Thanks!

VibhuJawa committed:
removed unused import

VibhuJawa commented:
> Thanks! Just added a small comment on an unused import.
Removed that. 
> Could you also retarget the PR to target branch-21.12? Thanks!

Have updated it to target `21.12`

ayushdg commented:
rerun tests

VibhuJawa committed:
Trigger Build

ayushdg commented:
rerun tests
